ERCS/erc-8003.md

```diff
@@ -0,0 +1,147 @@
+discussions-to: https://ethereum-magicians.org/t/erc-8003-erc-20-pre-initialization-extension-sentinel-storage/24993
+status: Draft
+type: Standards Track
+category: ERC
+created: 2025-07-17
+requires: 20
+---
+
+
+## Abstract
+
+This [ERC-20](../EIPS/eip-20.md) extension introduces an **optional pre-initialization function** for ERC-20 tokens, leveraging a sentinel (magic) value and a `bytes32` storage mapping for balances. This mechanism allows users to pay the high SSTORE gas cost in advance (when gas is cheap), so that the later first real token transfer to that address costs far less gas during a potential upcoming network congestion.
+
+Pre-initializing an ERC-20 token balance storage slot can save significant gas for users, but a simple approach (writing 0 to a `uint256` balance) does **not** actually reduce costs for the eventual first real transfer at all. This extension proposes a safe, ERC-20-compatible method using `mapping(address => bytes32)` for balances and a unique sentinel value. The contract's ERC-20 interface remains 100% standard: all reads/writes are still performed as `uint256`. Only the internal pre-initialization function ever stores the sentinel value; all normal functions interpret it as zero, and any later transfer/mint simply overwrites it.
+
+This allows users or integrators to "prepay" storage rent and optimize gas costs before high-demand launches, presales, and trending tokens. An easy way to understand for experienced developers is that this mirrors the principle of gas-shifting found in [ERC-721](./eip-721)A, but applied in reverse for ERC-20 storage.
+
+
+## Motivation
+
+In Ethereum, the **first write to a new storage slot** costs 20000 gas (SSTORE from zero to nonzero). Pre-initializing by setting zero to a `uint256` balance does *not* help, since the cost is incurred only when a nonzero value is first written. By using a sentinel (magic) value in a `bytes32` mapping, contracts can explicitly allocate the storage slot for a given address, paying the high gas fee in advance, and later updating the slot with the real balance at only 5000 gas.
+
+This makes gas costs predictable and manageable for scenarios where a spike in network fees is likely at time of launch, claim, or trending event.
+
+**Use cases (motivating examples):**
+1. **Presale Gas Hedging**
+   - During a presale or major launch, participants can pre-initialize their slot the night before, at low gas prices. Their eventual claim or buy is then cheaper.
+2. **Strategic Pre-initialization for Trending Tokens**
+   - Savvy users can batch-initialize slots on tokens they may want to trade when hype spikes.
+3. **Wallet/dApp Automation**
+   - Wallets, bots, and advanced dApps can offer a "pre-initialize now" button or automatic service for users to optimize their on-chain costs.
+
+
+## Specification
+
+ERC-20 contracts implementing this extension **MUST**:
+- Store balances as `mapping(address => bytes32)`.
```

Comment:
This is an implementation detail. If the contract wants to implement it using some other mechanism, but the observable external effects are the same, unless you're expecting people to directly read the storage, through, for example, `eth_getStorageAt`.